WebClient seems to have race condition

## Environment Details
* Helidon Version: 2.x
* Helidon SE or Helidon MP
* JDK version:
* OS:
* Docker version (if applicable):

----------

## Problem Description


## Steps to reproduce
1. Start the Helidon MP QuickStart app at port 8080.
2. Expand the attached zip [myproject.zip](https://github.com/oracle/helidon/files/7368816/myproject.zip)
3. `cd myproject`
4. `mvn clean test -Dtest=DefaultApiTest`

Running locally on my Mac, two of the three tests fail:
```
[INFO] Running io.helidon.examples.openapi.webclient.api.DefaultApiTest
[ERROR] Tests run: 3, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.827 s <<< FAILURE! - in io.helidon.examples.openapi.webclient.api.DefaultApiTest
[ERROR] getPersonalizedGreetingTest(io.helidon.examples.openapi.webclient.api.DefaultApiTest)  Time elapsed: 0.498 s  <<< ERROR!
io.helidon.examples.openapi.webclient.ApiException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Value supplier not yet set
	at io.helidon.examples.openapi.webclient.api.DefaultApiTest.getPersonalizedGreetingTest(DefaultApiTest.java:77)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Value supplier not yet set
	at io.helidon.examples.openapi.webclient.api.DefaultApiTest.getPersonalizedGreetingTest(DefaultApiTest.java:77)
Caused by: java.lang.IllegalStateException: Value supplier not yet set
	at io.helidon.examples.openapi.webclient.api.DefaultApiTest.getPersonalizedGreetingTest(DefaultApiTest.java:77)

```
Running the test methods individually works for all of them.

Now, comment out line 66 in `pom.xml` (the line specifying `<parallel>`). The tests run successfully, with the possible exception of getting 
```
org.junit.ComparisonFailure: Returned default greeting expected:<H[ello] World!> but was:<H[ola] World!>
```
because of test ordering. But the original problem is gone with the tests running serially instead of in parallel.

It seems running the tests  _in parallel_ exposes a race condition somewhere. I do not think it is in the example project.

```
Caused by: java.lang.IllegalStateException: Value supplier not yet set
	at io.helidon.webclient.NettyClient.lambda$static$1(NettyClient.java:75)
	at io.helidon.common.LazyValueImpl.get(LazyValueImpl.java:118)
	at io.helidon.webclient.WebClientRequestBuilderImpl.lambda$invoke$21(WebClientRequestBuilderImpl.java:571)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2333)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
	at io.helidon.webclient.WebClientRequestBuilderImpl.invoke(WebClientRequestBuilderImpl.java:543)
	at io.helidon.webclient.WebClientRequestBuilderImpl.lambda$request$13(WebClientRequestBuilderImpl.java:400)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:137)
	at io.helidon.webclient.WebClientRequestBuilderImpl.request(WebClientRequestBuilderImpl.java:400)
	at io.helidon.webclient.WebClientRequestBuilderImpl.submit(WebClientRequestBuilderImpl.java:405)
```

--- a/webclient/webclient/src/main/java/io/helidon/webclient/NettyClient.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/NettyClient.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2021 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -16,13 +16,14 @@
 package io.helidon.webclient;
 
 import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 import java.util.concurrent.ThreadFactory;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicReference;
 
 import io.helidon.common.LazyValue;
 import io.helidon.common.Version;
+import io.helidon.common.context.Contexts;
 import io.helidon.common.http.Http;
 import io.helidon.config.Config;
 import io.helidon.media.common.MediaContext;
@@ -36,7 +37,6 @@ import io.netty.channel.nio.NioEventLoopGroup;
  *       - what about the base URI? only would work with prod config
  */
 final class NettyClient implements WebClient {
-    private static final Config EMPTY_CONFIG = Config.empty();
     private static final Duration DEFAULT_CONNECT_TIMEOUT = Duration.ofMinutes(1);
     private static final Duration DEFAULT_READ_TIMEOUT = Duration.ofMinutes(10);
     private static final boolean DEFAULT_FOLLOW_REDIRECTS = false;
@@ -49,31 +49,54 @@ final class NettyClient implements WebClient {
     private static final MediaContext DEFAULT_MEDIA_SUPPORT = MediaContext.create();
     private static final WebClientTls DEFAULT_TLS = WebClientTls.builder().build();
 
-    private static final AtomicBoolean DEFAULTS_CONFIGURED = new AtomicBoolean();
-
-    private static final WebClientConfiguration DEFAULT_CONFIGURATION =
-            WebClientConfiguration.builder()
-                    .config(EMPTY_CONFIG)
-                    .connectTimeout(DEFAULT_CONNECT_TIMEOUT)
-                    .readTimeout(DEFAULT_READ_TIMEOUT)
-                    .followRedirects(DEFAULT_FOLLOW_REDIRECTS)
-                    .maxRedirects(DEFAULT_NUMBER_OF_REDIRECTS)
-                    .userAgent(DEFAULT_USER_AGENT)
-                    .readerContextParent(DEFAULT_MEDIA_SUPPORT.readerContext())
-                    .writerContextParent(DEFAULT_MEDIA_SUPPORT.writerContext())
-                    .proxy(DEFAULT_PROXY)
-                    .tls(DEFAULT_TLS)
-                    .keepAlive(DEFAULT_KEEP_ALIVE)
-                    .validateHeaders(DEFAULT_VALIDATE_HEADERS)
-                    .build();
-
     // configurable per client instance
-    static final AtomicReference<WebClientConfiguration> SHARED_CONFIGURATION = new AtomicReference<>(DEFAULT_CONFIGURATION);
+    static final WebClientConfiguration SHARED_CONFIGURATION;
 
     // shared by all client instances
-    private static LazyValue<NioEventLoopGroup> eventGroup = LazyValue.create(() -> {
-        throw new IllegalStateException("Value supplier not yet set");
-    });
+    private static final NioEventLoopGroup EVENT_GROUP;
+
+    static {
+        Config globalConfig = Contexts.globalContext().get(Config.class).orElseGet(Config::empty);
+
+        Config config = globalConfig.get("client");
+
+        WebClientConfiguration.Builder<?, ?> builder = WebClientConfiguration.builder()
+                .connectTimeout(DEFAULT_CONNECT_TIMEOUT)
+                .readTimeout(DEFAULT_READ_TIMEOUT)
+                .followRedirects(DEFAULT_FOLLOW_REDIRECTS)
+                .maxRedirects(DEFAULT_NUMBER_OF_REDIRECTS)
+                .userAgent(DEFAULT_USER_AGENT)
+                .readerContextParent(DEFAULT_MEDIA_SUPPORT.readerContext())
+                .writerContextParent(DEFAULT_MEDIA_SUPPORT.writerContext())
+                .proxy(DEFAULT_PROXY)
+                .tls(DEFAULT_TLS)
+                .keepAlive(DEFAULT_KEEP_ALIVE)
+                .validateHeaders(DEFAULT_VALIDATE_HEADERS)
+                .config(config);
+
+        SHARED_CONFIGURATION = builder.build();
+
+        Config eventLoopConfig = config.get("event-loop");
+        int numberOfThreads = eventLoopConfig.get("workers")
+                .asInt()
+                .orElse(1);
+        String threadNamePrefix = eventLoopConfig.get("name-prefix")
+                .asString()
+                .orElse("helidon-client-");
+        AtomicInteger threadCounter = new AtomicInteger();
+
+        ThreadFactory threadFactory =
+                r -> {
+                    Thread result = new Thread(r, threadNamePrefix + threadCounter.getAndIncrement());
+                    // we should exit the VM if client event loop is the only thread(s) running
+                    result.setDaemon(true);
+                    return result;
+                };
+
+        ExecutorService executorService = Executors.newCachedThreadPool(threadFactory);
+
+        EVENT_GROUP = new NioEventLoopGroup(numberOfThreads, Contexts.wrap(executorService));
+    }
 
     // this instance configuration
     private final WebClientConfiguration configuration;
@@ -88,11 +111,11 @@ final class NettyClient implements WebClient {
 
         // we need to configure these - if user wants to override, they must
         // do it before first usage
-        configureDefaults(EMPTY_CONFIG);
+//        configureDefaults(EMPTY_CONFIG);
     }
 
-    static LazyValue<NioEventLoopGroup> eventGroup() {
-        return eventGroup;
+    static NioEventLoopGroup eventGroup() {
+        return EVENT_GROUP;
     }
 
     @Override
@@ -132,41 +155,12 @@ final class NettyClient implements WebClient {
 
     @Override
     public WebClientRequestBuilder method(String method) {
-        return WebClientRequestBuilderImpl.create(eventGroup, configuration, Http.RequestMethod.create(method));
+        return WebClientRequestBuilderImpl.create(EVENT_GROUP, configuration, Http.RequestMethod.create(method));
     }
 
     @Override
     public WebClientRequestBuilder method(Http.RequestMethod method) {
-        return WebClientRequestBuilderImpl.create(eventGroup, configuration, method);
-    }
-
-    static void configureDefaults(Config globalConfig) {
-        if (DEFAULTS_CONFIGURED.compareAndSet(false, true)) {
-            Config config = globalConfig.get("client");
-            WebClientConfiguration.Builder<?, ?> builder = DEFAULT_CONFIGURATION.derive();
-            Config eventLoopConfig = config.get("event-loop");
-            int numberOfThreads = eventLoopConfig.get("workers")
-                    .asInt()
-                    .orElse(1);
-            String threadNamePrefix = eventLoopConfig.get("name-prefix")
-                    .asString()
-                    .orElse("helidon-client-");
-            AtomicInteger threadCounter = new AtomicInteger();
-
-            ThreadFactory threadFactory =
-                    r -> {
-                        Thread result = new Thread(r, threadNamePrefix + threadCounter.getAndIncrement());
-                        // we should exit the VM if client event loop is the only thread(s) running
-                        result.setDaemon(true);
-                        return result;
-                    };
-
-            eventGroup = LazyValue.create(new NioEventLoopGroup(numberOfThreads, threadFactory));
-
-            builder.config(config);
-
-            SHARED_CONFIGURATION.set(builder.build());
-        }
+        return WebClientRequestBuilderImpl.create(EVENT_GROUP, configuration, method);
     }
 
 }

--- a/webclient/webclient/src/main/java/io/helidon/webclient/NettyClient.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/NettyClient.java
@@ -49,18 +49,17 @@ final class NettyClient implements WebClient {
     private static final MediaContext DEFAULT_MEDIA_SUPPORT = MediaContext.create();
     private static final WebClientTls DEFAULT_TLS = WebClientTls.builder().build();
 
+    private static final Config GLOBAL_CLIENT_CONFIG;
+
     // configurable per client instance
     static final WebClientConfiguration SHARED_CONFIGURATION;
-
-    // shared by all client instances
-    private static final NioEventLoopGroup EVENT_GROUP;
 
     static {
         Config globalConfig = Contexts.globalContext().get(Config.class).orElseGet(Config::empty);
 
-        Config config = globalConfig.get("client");
+        GLOBAL_CLIENT_CONFIG = globalConfig.get("client");
 
-        WebClientConfiguration.Builder<?, ?> builder = WebClientConfiguration.builder()
+        SHARED_CONFIGURATION = WebClientConfiguration.builder()
                 .connectTimeout(DEFAULT_CONNECT_TIMEOUT)
                 .readTimeout(DEFAULT_READ_TIMEOUT)
                 .followRedirects(DEFAULT_FOLLOW_REDIRECTS)
@@ -72,11 +71,13 @@ final class NettyClient implements WebClient {
                 .tls(DEFAULT_TLS)
                 .keepAlive(DEFAULT_KEEP_ALIVE)
                 .validateHeaders(DEFAULT_VALIDATE_HEADERS)
-                .config(config);
+                .config(GLOBAL_CLIENT_CONFIG)
+                .build();
+    }
 
-        SHARED_CONFIGURATION = builder.build();
-
-        Config eventLoopConfig = config.get("event-loop");
+    // shared by all client instances
+    private static final LazyValue<NioEventLoopGroup> EVENT_GROUP = LazyValue.create(() -> {
+        Config eventLoopConfig = GLOBAL_CLIENT_CONFIG.get("event-loop");
         int numberOfThreads = eventLoopConfig.get("workers")
                 .asInt()
                 .orElse(1);
@@ -95,8 +96,8 @@ final class NettyClient implements WebClient {
 
         ExecutorService executorService = Executors.newCachedThreadPool(threadFactory);
 
-        EVENT_GROUP = new NioEventLoopGroup(numberOfThreads, Contexts.wrap(executorService));
-    }
+        return new NioEventLoopGroup(numberOfThreads, Contexts.wrap(executorService));
+    });
 
     // this instance configuration
     private final WebClientConfiguration configuration;
@@ -115,7 +116,7 @@ final class NettyClient implements WebClient {
     }
 
     static NioEventLoopGroup eventGroup() {
-        return EVENT_GROUP;
+        return EVENT_GROUP.get();
     }
 
     @Override
@@ -155,12 +156,12 @@ final class NettyClient implements WebClient {
 
     @Override
     public WebClientRequestBuilder method(String method) {
-        return WebClientRequestBuilderImpl.create(EVENT_GROUP, configuration, Http.RequestMethod.create(method));
+        return WebClientRequestBuilderImpl.create(EVENT_GROUP.get(), configuration, Http.RequestMethod.create(method));
     }
 
     @Override
     public WebClientRequestBuilder method(Http.RequestMethod method) {
-        return WebClientRequestBuilderImpl.create(EVENT_GROUP, configuration, method);
+        return WebClientRequestBuilderImpl.create(EVENT_GROUP.get(), configuration, method);
     }
 
 }

--- a/webclient/webclient/src/main/java/io/helidon/webclient/WebClient.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/WebClient.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2021 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2022 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -146,7 +146,7 @@ public interface WebClient {
                                    ParentingMediaContextBuilder<Builder>,
                                    MediaContextBuilder<Builder> {
 
-        private final WebClientConfiguration.Builder<?, ?> configuration = NettyClient.SHARED_CONFIGURATION.get().derive();
+        private final WebClientConfiguration.Builder<?, ?> configuration = NettyClient.SHARED_CONFIGURATION.derive();
         private final HelidonServiceLoader.Builder<WebClientServiceProvider> services = HelidonServiceLoader
                 .builder(ServiceLoader.load(WebClientServiceProvider.class));
         private final List<WebClientService> webClientServices = new ArrayList<>();

--- a/webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java
@@ -42,7 +42,6 @@ import java.util.function.Function;
 import java.util.logging.Logger;
 
 import io.helidon.common.GenericType;
-import io.helidon.common.LazyValue;
 import io.helidon.common.context.Context;
 import io.helidon.common.context.Contexts;
 import io.helidon.common.http.DataChunk;
@@ -62,7 +61,6 @@ import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelOption;
-import io.netty.channel.EventLoopGroup;
 import io.netty.channel.nio.NioEventLoopGroup;
 import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.handler.codec.http.DefaultHttpHeaders;
@@ -110,7 +108,7 @@ class WebClientRequestBuilderImpl implements WebClientRequestBuilder {
     }
 
     private final Map<String, String> properties;
-    private final LazyValue<NioEventLoopGroup> eventGroup;
+    private final NioEventLoopGroup eventGroup;
     private final WebClientConfiguration configuration;
     private final Http.RequestMethod method;
     private final WebClientRequestHeaders headers;
@@ -136,7 +134,7 @@ class WebClientRequestBuilderImpl implements WebClientRequestBuilder {
     private Long requestId;
     private boolean allowChunkedEncoding;
 
-    private WebClientRequestBuilderImpl(LazyValue<NioEventLoopGroup> eventGroup,
+    private WebClientRequestBuilderImpl(NioEventLoopGroup eventGroup,
                                         WebClientConfiguration configuration,
                                         Http.RequestMethod method) {
         this.properties = new HashMap<>();
@@ -167,9 +165,9 @@ class WebClientRequestBuilderImpl implements WebClientRequestBuilder {
         this.keepAlive = configuration.keepAlive();
     }
 
-    public static WebClientRequestBuilder create(LazyValue<NioEventLoopGroup> eventGroup,
-                                                 WebClientConfiguration configuration,
-                                                 Http.RequestMethod method) {
+    static WebClientRequestBuilder create(NioEventLoopGroup eventGroup,
+                                          WebClientConfiguration configuration,
+                                          Http.RequestMethod method) {
         return new WebClientRequestBuilderImpl(eventGroup, configuration, method);
     }
 
@@ -568,9 +566,8 @@ class WebClientRequestBuilderImpl implements WebClientRequestBuilder {
 
             CompletableFuture<WebClientResponse> result = new CompletableFuture<>();
 
-            EventLoopGroup group = eventGroup.get();
             Bootstrap bootstrap = new Bootstrap();
-            bootstrap.group(group)
+            bootstrap.group(eventGroup)
                     .channel(NioSocketChannel.class)
                     .handler(new NettyClientInitializer(requestConfiguration))
                     .option(ChannelOption.SO_KEEPALIVE, keepAlive)

